A51-custom-backend-metrics.md

```diff
@@ -0,0 +1,387 @@
+
+```Java
+class CustomPicker extends SubchannelPicker {
+  public PickResult pickSubchannel(PickSubchannelArgs args) {
+  Subchannel subchannel = … // WRR picking logic
+  return PickResult.withSubchannel(
+    subchannel,
+    OrcaUtil.newOrcaClientStreamTracerFactory(listener));
+  }
+}
+```
+
+#### Out-of-Band Reporting 
+At the server side, Java implemented the OOB service `OrcaServiceImpl`: users can optionally provide the minimum
+reporting interval configuration when creating the object. Then they can update utilization metrics
+data using the exposed APIs as follows. All the methods are thread safe.
+
+```java
+public class OrcaServiceImpl implements OpenRcaService {
+
+// Allow configuring minimum report interval. If not configured or badly configured non-positive, 
+// the default is 30s.
+// In the future, it might be useful to let users specify minimum report interval and default 
+// report interval separately. If the default report interval is not configured or badly configured 
+// non-positive, the default is 1 min. 
+// public OpenRcaServiceImpl(long minInterval, TimeUnit timeUnit, long 
+// defaultInterval, TimeUnit)
+public OrcaServiceImpl(long minReportInterval, TimeUnit unit);
+
+// Creates an OOB service using default minimum report interval = 30s.
+public OrcaServiceImpl();
+
+// Update the metrics value corresponding to the specified key.
+public void setUtilizationMetric(String key, double value);
+
+// Replace the whole metrics data using the specified map.
+public void setAllUtilizationMetrics(Map<String, Double> metrics);
+
+// Remove the metrics data entry corresponding to the specified key.
+public void deleteUtilizationMetric(String key);
+
+// Update the CPU utilization metrics data.
+public void setCPUUtilizationMetric(double value);
+
+// Clear the CPU utilization metrics data.
+public void deleteCPUUtilizationMetric();
+
+// Update the memory utilization metrics data.
+public void setMemoryUtilizationMetric(double value);
+
+// Clear the memory utilization metrics data.
+public void deleteMemoryUtilizationMetric();
+}
+```
+
+At the client side, Java will provide a utility function to wrap the `LoadBalancer.Helper`,
+usable by both WRR policy and xDS LB policies. The `LoadBalancer.Helper` manages the OOB stream
+lifecycle and accepts subscriptions from any routing hierarchy.
+The OOB stream will also gracefully disable itself if the server integration is not complete, 
+so users do not have to worry about client and server deployment coordination.
+
+The following pseudocode shows how a custom load balancing policy instantiates the helper wrapper, 
+installs the out-of-band metrics reporting mechanism, and then consumes the metrics reports, 
+which only requires users to register an `OrcaOobReportListener` that contains their own business logic:
+
+```java
+class CustomLoadBalancer extends LoadBalancer {
+  private final Helper helper;  // the original Helper
+  void handleNewServer(address) {
```

Comment:
It isn't clear to me what this function's purpose is. It creates stuff, but then throws is away.